core, cmd/puppeth: implement constantinople fix, disable EIP-1283

holiman opened the pull request:
This PR adds a new fork, which disables EIP-1283. Internally it's called `Petersburg`, but the genesis/config field is `ConstantinopleFix`. 

- This PR also contains code for puppeth to generate chainspec for Parity, using `eip1283DisableTransition`, which _appear_ to be what they are going with (see https://github.com/paritytech/parity-ethereum/pull/10214 ). 
- The blocknumbers used in this pr are
  - `7280000` for Constantinople (Mainnet)
  - `7280000`  for ConstantinopleFix (Mainnet) 
  - `4939394` for ConstantinopleFix on Ropsten
  - `9999999` for ConstantinopleFix on Rinkeby (todo decide on a real number)

This PR also defaults to using the same `ConstantinopleFix` number as whatever `Constantinople` is set to. That is, it will default to mainnet behaviour if `ConstantinopleFix` is `nil`. 

The puppeth generator will allow setting to disable `eip 1283` as early as block `0` -- i.e, it does not automatically enable the rest of constantinople. 

**OBS** This means that for private networks, which have already transitioned to Constantinople, this PR _will_ break the network unless `ConstantinopleFix` is explicitly set!  -- becuase if unset, it will default to same number as `Constantinople`. This is a usability tradeoff, but I figured that defaulting to mainnet behaviour is the Right Thing To Do (tm)

5chdn commented on cmd/puppeth/genesis.go:
```diff
@@ -441,6 +447,14 @@ func (spec *parityChainSpec) setConstantinople(num *big.Int) {
+	spec.Params.EIP145Transition = n
+	spec.Params.EIP1014Transition = n
+	spec.Params.EIP1052Transition = n
+	spec.Params.EIP1283Transition = n
```
here `EIP1283DisableTransition` ?

holiman replied:
Thanks

5chdn commented on params/config.go:
```diff
@@ -91,6 +93,7 @@ var (
 		EIP158Block:         big.NewInt(3),
 		ByzantiumBlock:      big.NewInt(1035301),
 		ConstantinopleBlock: big.NewInt(3660663),
+		PetersburgBlock:     big.NewInt(9999999), //TODO! Insert Rinkeby block number
```
What about `3800083`?

holiman replied:
Now we're at `3729555`, `3800083` is in abput `11.4` days from now. @karalabe what's your take?

5chdn commented:
created an EIP for that https://github.com/ethereum/EIPs/pull/1716

holiman commented:
I updated the tests to use the newly generated tests from https://github.com/ethereum/tests/pull/582 . 
I also updated the output to be less verbose when tests are failing. The previous op-by-op output where every op takes about 40 lines is totally useless, so now it uses the new `json` based format instead, to make it easier to investigate test failures.

holiman commented:
Seems there are only unrelated errors remaining on travis. @fjl  PTAL
